libzfs_core/_libzfs_core.py

```diff
@@ -663,6 +668,419 @@ def lzc_exists(name):
+        os.close(fd)
+
+    if entry is None:
+        raise exceptions.DatasetNotFound(name)
+    result = result['properties']
+    # In most cases the source of the property is uninteresting and the
+    # value alone is sufficient.  One exception is the 'mountpoint'
+    # property the final value of which is not the same as the inherited
+    # value.
+    mountpoint = result.get('mountpoint')
+    if mountpoint is not None:
+        mountpoint_src = mountpoint['source']
+        mountpoint_val = mountpoint['value']
+        # 'source' is the name of the dataset that has 'mountpoint' set
+        # to a non-default value and from which the current dataset inherits
+        # the property.  'source' can be the current dataset if its
+        # 'mountpoint' is explicitly set.
+        # 'source' can also be a special value like '$recvd', that case
+        # is equivalent to the property being set on the current dataset.
+        if not mountpoint_src.startswith('$'):
+            mountpoint_val = mountpoint_val + name[len(mountpoint_src):]
+    else:
+        mountpoint_val = '/' + name
+    result = { k: v['value'] for k, v in result.iteritems() }
+    if 'clones' in result:
+        result['clones'] = [ x for x in result['clones'] ]
+    result['mountpoint'] = mountpoint_val
+    return result
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
+    '''
+    List the children of the ZFS dataset.
+
+    :param bytes name: the name of the dataset.
+    :return: an iterator that produces the names of the children.
+    :raises NameInvalid: if the dataset name is invalid.
+    :raises NameTooLong: if the dataset name is too long.
+
+    .. warning::
+        If the dataset does not exist, then the returned iterator would produce
+        no results and no error is reported.
+        That case is indistinguishable from the dataset having no children.
+
+        An attempt to list children of a snapshot is silently ignored as well.
+    '''
+    children = []
+    (fd, other_fd) = lzc_list(name, {'recurse': 1, 'type': {'filesystem': None, 'volume': None}})
+    try:
+        while True:
+            record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+            if not record_bytes:
+                break
+            (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+            xlate.lzc_list_children_xlate_error(err, name)
+            if size == 0:
+                break
+            data_bytes = os.read(fd, size)
+            result = {}
+            with nvlist_out(result) as nvp:
+                ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+            if ret != 0:
+                raise ZFSError(ret, "Failed to unpack list data")
+            child = result['name']
+            if child != name:
+                children.append(child)
+    finally:
+        os.close(other_fd)
+        os.close(fd)
+
+    def _iterator():
+        while children:
+            yield children.pop(0)
+
+    return _iterator()
```

Comment:
I realize I suggested the generator-style approach that finishes off this function.  I think the `iter(children)` solution is probably just as good.  Initially I had thought there would only be I/O done in the first loop and most of the parsing would happen afterwards in this generator.  Considering some parsing has to be done as the I/O is happening (to extract the lengths to know how much more to read), that approach probably wouldn't provide any benefit after all.